softlayer/resource_softlayer_subnet.go

```
@@ -0,0 +1,386 @@
+				ConflictsWith: []string{"vlan_id"},
+			},
+
+			// Provides IP address/netmask format (ex. 10.10.10.10/28)
+			"subnet_cidr": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"notes": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceSoftLayerSubnetCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ProviderConfig).SoftLayerSession()
+
+	// Find price items with AdditionalServicesSubnetAddresses
+	productOrderContainer, err := buildSubnetProductOrderContainer(d, sess)
+	if err != nil {
+		return fmt.Errorf("Error creating subnet: %s", err)
+	}
+
+	log.Println("[INFO] Creating subnet")
+
+	receipt, err := services.GetProductOrderService(sess).
+		PlaceOrder(productOrderContainer, sl.Bool(false))
+	if err != nil {
+		return fmt.Errorf("Error during creation of subnet: %s", err)
+	}
+
+	Subnet, err := findSubnetByOrderId(sess, *receipt.OrderId)
+	if err != nil {
+		return fmt.Errorf("Error during creation of subnet: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%d", *Subnet.Id))
+
+	return resourceSoftLayerSubnetUpdate(d, meta)
+}
+
+func resourceSoftLayerSubnetRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ProviderConfig).SoftLayerSession()
+	service := services.GetNetworkSubnetService(sess)
+
+	subnetId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid subnet ID, must be an integer: %s", err)
+	}
+
+	subnet, err := service.Id(subnetId).Mask(SubnetMask).GetObject()
+	if err != nil {
+		return fmt.Errorf("Error retrieving a subnet: %s", err)
+	}
+
+	if *subnet.AddressSpace == "PRIVATE" {
+		d.Set("private", true)
+	} else if *subnet.AddressSpace == "PUBLIC" {
+		d.Set("private", false)
+	}
+
+	d.Set("type", *subnet.SubnetType)
+	if strings.Contains(*subnet.SubnetType, "STATIC") {
+		d.Set("type", "Static")
+	} else if strings.Contains(*subnet.SubnetType, "VLAN") {
+		d.Set("type", "Portable")
+	}
```

@minsikl There are `SubnetType`s that would not contain STATIC or VLAN. In that case "type" would be unset. Is that intentional?

---

@renier Portable subnets contain "VLAN" and Static subnets contain "STATIC". In the future, additional types can be added. For example, Primary subnets contain "PRIMARY".

---

@minsikl Since there are some types that do not contain either word, technically, would it would be possible to get that from the API? If so, programmatically, it would be safer to add an else clause to handle that situation. For example, to throw an error in that case, since we don't support any other type of subnet.

---

@renier Users may try to import different type of subnets. The current code will not generate an error. If you want to check the type strictly, you can throw an exception.

---

Updated the code to generate an error when an invalid subnet type is detected.